crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1673,6 +1673,128 @@ impl<'db> Type<'db> {
+                // = (A & Falsy) & ~B
+                // ≈ A.exclude_always_truthy() & ~B
+                //
+                // Case 2: Positives are empty
+                // ~B & Falsy
+                // = ~(B | AlwaysTruthy)
+                // ≈ ~(B.include_always_truthy())
+                //
+                // Note: Calculating `(~B & Falsy)` = `~(B | AlwaysTruthy)` is extremely difficult.
+                // This set should include all elements where `bool(element)` is always True,
+                // regardless of the type of B. Examples include:
+                // { *(elements of B), True, 1, 2, ..., -1, -2, ..., *(instances of object), ... }.
+                //
+                // Attempting to represent this set via Union or Intersection
+                // could break type inference.
+                let mut builder = IntersectionBuilder::new(db);
+
+                // Theoretically, applying `intersection` to a single positive or to all positives
+                // should yield the same result. However, due to the limitations of the current type system,
+                // there is a possibility of losing information when applying it to only one positive.
+                // As a result, `exclude_always_truthy` is applied to all positives.
+                for element in intersection.positive(db) {
+                    builder = builder.add_positive(element.exclude_always_truthy(db));
+                }
+
+                for element in intersection.negative(db) {
+                    builder = builder.add_negative(*element);
+                }
+
+                debug_assert!(
+                    !intersection.positive(db).is_empty(),
+                    "negative-only-intersections can not use `exclude_always_truthy`"
+                );
+
+                builder.build()
+            }
+            Type::Union(union) => {
+                let mut builder = UnionBuilder::new(db);
+
+                for element in union.elements(db) {
+                    builder = builder.add(element.exclude_always_truthy(db));
+                }
+
+                builder.build()
+            }
+            // Since `bool` is a @final class, there is no need to worry about subclasses.
+            Type::Instance(instance) if instance.class.is_known(db, KnownClass::Bool) => {
+                Type::BooleanLiteral(false)
+            }
+            _ => {
+                if let Truthiness::AlwaysTrue = self.bool(db) {
+                    return Type::Never;
+                }
+
+                self
+            }
+        }
+    }
+
+    pub fn falsy_literals(db: &'db dyn Db) -> Type<'db> {
+        UnionBuilder::new(db)
+            .add(Type::none(db))
+            .add(Type::BooleanLiteral(false))
+            .add(Type::IntLiteral(0))
+            .add(Type::string_literal(db, ""))
+            .add(Type::bytes_literal(db, b""))
+            // TODO: tuple literal should be included
+            // .add(Type::tuple(db, &[]))
```

Comment:
I’m not sure if this is the correct way to represent `()`.  
It also seems like `()` hasn’t been considered yet in `is_disjoint_from`.  

I think it make sense to implement support for `()` first. Adding this code as-is makes the mdtest results much harder to read.

Reply:
> I’m not sure if this is the correct way to represent `()`.

I think it is!

> It also seems like `()` hasn’t been considered yet in `is_disjoint_from`.

That's very possible

Reply:
Got it! Ah, my point was that the current implementation without TupleLiteral produces clean results like `A | B`. However, if we uncomment the TupleLiteral, it results like `A & ~Tuple[()] | B & ~Tuple[()]`, which makes the tests harder to read.

I’ll try to resolve this in a different PR before we uncomment Tuple Literal.